### PR TITLE
Add selection by array

### LINKED ISSUE #416: Support field selection with complex keys
When passing in columns, I'd like to be able to select data from an object when the keys in that object include full-stops.

For example, here is an entity from a KRM with 'annotations':

```js
{ entity: { metadata: { annotations: { "example.com/custom-1": "foo"}}}
```

in this case I might want to try using a `field` attribute of `entity.metadata.annotations["example.com/custom-1"]`.

This syntax works for  numbers (e.g. `a[1]`), and KIND OF works for strings, as long as it doesn't contain full-stops.

This appears to come down to an implementation for 'byString' which first converts array syntax (`a['key']`) to object notation (`a.key`) before processing here:

https://github.com/material-table-core/core/blob/089917160b200179d9b2ace6e65c3402f74ad08a/src/utils/index.js#L7

## Background

I'm using a data Table through another tool, specifically [Backstage](https://github.com/backstage/backstage). And Kubernetes Resource Models (KRM) is the standard notation for defining entities. This leads to objects like in the above example.

## Suggested solution

I think there are two solutions, either:
 1. make field work for strings with array syntax with full-stops in keys
 2. allow an array to be passed in, that explicitly declares the keys

Whilst the first option is possible it has the potential to change behaviour for existing consumers (not that I can think of a scenario where a consumer would want the current behaviour if they have a key with a full-stop).

So I'd suggest (and have a PR ready) to implement the latter. But also happy to see this added in any way, and if not I've found a hack solution to work around (by passing in an object with an overloaded 'split' method).

## Thanks

Thank-you for considering this issue. I'll follow up with a PR shortly.

### PR DESCRIPTION
**Pull request by jahredhope**
## Related Issue

Closes https://github.com/material-table-core/core/issues/416

## Description

Support passing an array to the `field` attribute of a column. Allowing the selection of keys with full-stops.

Details for why this would be valuable can be found in the [related issue](https://github.com/material-table-core/core/issues/416).

## Related PRs

None that I can find.

## Impacted Areas in Application

- Table `columns` prop

## Additional Notes

Thanks for considering this PR and your hard work on this project.

### CODE REVIEW

**Comment by matthewoestreich** [comment]
Thank you for this! Looks good too me - just want to get @Domino987 approval on this before merging.

**Review by Domino987 — CHANGES_REQUESTED** [review]
This looks awesome and thanks for this.
Also tests, nice :)
Just one small request, would you mind updating the type of field in the types file to match this as well?

**Comment by jahredhope** [comment]
Ah yes, sorry about missing that. I thought the types came in from DefinitelyTyped. As someone who uses mostly TypeScript I appreciate you ensuring they are kept up-to-date.

Updated.

**Review by Domino987 — APPROVED** [review]

**Pull request merged**
